### ISSUE TITLE
Handle Location/EVSE lifecycle more clearly

### ISSUE THREAD

**Issue by julius-yang** [ocpi/ocpi#412]
The currently released versions of OCPI require that a Location or EVSE remain in the CPO's OCPI data forever.  A REMOVED status for an EVSE indicates EVSE decommissioning, and a Location where all the EVSEs have a REMOVED status indicates Location decommissioning.

However in real world usage CPOs routinely ignore this part of the specification (or did not realize that it was a requirement).  Much of it boils down to operational reasoning.  Here are two examples PlugShare (an eMSP) has seen which I consider legitimate use cases for no longer showing an EVSE.

1. CPO accidentally exposes EVSEs and hides them again (for varied reasons, e.g. internal mistake or site operator does not want their Location represented to eMSPs)
2. CPO sells chargers to another entity.  It seems unreasonable that the CPO should represent these EVSEs in perpetuity.

Given that so many CPOs ignore this part of the spec because of real-world considerations, I think the spec should allow data to vanish and moreover clearly define what it means when the data vanishes. Ideally those definitions would allow a per-CPO or even per-entity way of defining what it means if that item vanishes.  I have some ideas but before proposing them I'm interested in knowing whether this change is even acceptable or possible.

PlugShare deals with dozens of OCPI services and understanding what happens when EVSEs/Locations vanish from that CPO's data is extremely laborious because it requires us to ask each CPO what it means.  Almost all of the services do it.

**Comment by reinierl** [comment]
Thanks Julius! I recognise your problems and I'll try to incorporate solutions for them in the OCPI 3.0 development process.

**Comment by julius-yang** [comment]
Thanks.  The simplest solution would be a DELETE request from the CPO to the eMSP. Though this still doesn't tell the eMSP what to do with the missing data, it does implicitly tell the eMSP that this data should no longer be shown by the eMSP.

My more complicated idea was to add properties to Location and EVSE classes with an enum indicating what should happen if that entity goes missing.

**Workflow**
1. CPO pushes or eMSP pulls a Location/EVSE
2. Location/EVSE has property `if_missing` (naming is hard)
3. `if_missing` contains enums describing actions the eMSP should take if the Location/EVSE no longer exists in the CPO's data, e.g. HIDE, DELETE, MARK_AS_UNDER_REPAIR, MARK_AS_COMING_SOON
4. CPO removes Location/EVSE from data
5. eMSP sees Location/EVSE is missing and looks at its own internal copy of the item and uses `is_missing` to take the appropriate action on its side (stop showing PlugShare item, but retain it; delete OCPI item, mark PlugShare Station as under repair, mark PlugShare Location as coming soon...)

By default, `if_missing` would have the value DELETE (unless CPOs and eMSPs agree some other process should be the default).  This allows eMSPs to wipe the data without needing to query the CPO.  If the CPO complains that we are no longer showing their data, we can point to the spec :)

I am definitely interested in hearing about other issues eMSPs and CPOs have encountered.  I don't want this to be a solution that only fits PlugShare.

**Comment by Alan-Stewart** [comment]
I thought the reason to keep these around forever is so that the billing info would always be correct as of the date of the charging event?

**Comment by julius-yang** [comment]
That's a really good point.  In our case for most CPO partners we only read Location and Tariff data, so the relationship of Location to CDRs and Sessions would break if the Location/EVSE were wiped.  I forgot about that.  Definitely makes things harder, but I still hope that handling deprecated data will be possible someday.

**Comment by reinierl** [comment]
All objects except CDRs can be mutated though so if the receiver party doesn't make copies of historic versions they can't be sure they have the correct info for billing anyway, can they?

**Comment by julius-yang** [comment]
@mathieu-muller Yeah my idea is not great.  I'm hoping to spark a discussion about EVSE lifecycle handling.  Sometimes you propose a terrible idea just to get things going, because engineers like to criticize things and improve them :)

Internally we distinguish between EVSEs that are explicitly marked REMOVED vs ones that vanish from the location data.  But the core problem remains: an EVSE that vanishes from the data or is marked REMOVED can mean different things to different CPOs.
